src/lfo.rs

```diff
@@ -0,0 +1,115 @@
+// implements a wavetable LFO
+
+
+use std::f32::consts::PI;
+use crate::ring_buffer::RingBuffer;
+
+/// LFO is a struct that contains a wavetable and a phase, frequency, and amplitude.
+pub struct LFO {
+    sample_rate: f32,
+    wavetable: RingBuffer<f32>,
+    phase: f32,
+    frequency: f32,
+    amplitude: f32,
+
+}
+/// implements functions for LFO struct
+impl LFO {
+    /// creates a new LFO with a given sample rate, frequency, and amplitude, only does so for one period
+    pub fn new(sample_rate: f32, frequency: f32, amplitude: f32) -> LFO {
+        let wavetable_size = (sample_rate / frequency) as usize;
+        let mut wavetable = RingBuffer::new(wavetable_size);
+        for i in 0..wavetable_size {
+            let value = amplitude * (2.0 * PI * i as f32 / wavetable_size as f32).sin();
+            wavetable.push(value);
+        }
+        LFO {
+            sample_rate,
+            wavetable,
+            phase: 0.0,
+            frequency,
+            amplitude: 1.0,
+        }
+    }
+
+    /// set a new frequency for the LFO
+    pub fn set_frequency(&mut self, frequency: f32) {
+        self.frequency = frequency;
+        let wavetable_size = (self.sample_rate / frequency) as usize;
+        self.wavetable = RingBuffer::new(wavetable_size);
+        for i in 0..wavetable_size {
+            let value = self.amplitude * (2.0 * PI * i as f32 / wavetable_size as f32).sin();
+            self.wavetable.push(value);
+        }
+    }
+    /// set a new amplitude for the LFO
+    pub fn set_amplitude(&mut self, amplitude: f32) {
+        self.amplitude = amplitude;
+        let wavetable_size = (self.sample_rate / self.frequency) as usize;
+        self.wavetable = RingBuffer::new(wavetable_size);
+        for i in 0..wavetable_size {
```

Comment:
You could do your amplitude multiplication in the get_sample function like this:
value = self.wavetable.peek() * self.amplitude;
This way you don't have to regenerate the entire waveform.

Reply:
That's a really good point!